packages/datagrid/src/components/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,156 @@
+import { GridContext, HeaderClickParams } from '../../types';
+import { handleKeyboard, handleMultiSelection } from './DataGrid.utils';
+
+import theme from './DataGrid.module.scss';
+
+export type DataGridProps = GridOptions & {
+	/* Shows loader icons instead of grid */
+	loading?: boolean;
+	/* Enable ctrl/shift modifier on column header, default: 'single' */
+	columnSelection?: 'single' | 'multiple';
+	onColumnSelectionChanged?(params: { columns: string[] }): void;
+	/* Controlled selection, handling only columns for now */
+	selection?: {
+		columnIds?: string[];
+	};
+};
+
+export default function DataGrid({
+	loading,
+	columnSelection = 'single',
+	onColumnSelectionChanged,
+	columnDefs,
+	selection,
+	...props
+}: DataGridProps): JSX.Element {
+	const gridRef = useRef<AgGridReact>(null);
+	const [selectedColumns, setSelectedColumns] = useState<string[]>([]);
+	const context = useRef<GridContext>({ selectedColumns });
+
+	const onCellFocused = useCallback(
+		(event: CellFocusedEvent) => {
+			// filter event triggered by clearFocusedCell
+			if (event.column instanceof Column) {
+				const update = event.column.isPinned() ? [] : [event.column.getColId()];
+				setSelectedColumns(prev => {
+					// don't update when focusing a cell on the same column
+					return prev.length > 1 || prev[0] !== update[0] ? update : prev;
+				});
+			}
+			if (props.onCellFocused) {
+				props.onCellFocused(event);
+			}
+		},
+		[props.onCellFocused, setSelectedColumns],
+	);
+
+	const onHeaderFocus = useCallback(
+		(event: HeaderClickParams) => {
+			// Unselect rows if needed
+			if (gridRef.current?.api.getSelectedRows().length) {
+				gridRef.current?.api.deselectAll();
+				gridRef.current?.api.clearFocusedCell();
+			}
+			const colId = event.column.getColId();
+			const allColIds = gridRef.current!.columnApi.getColumns()!.map(col => col.getColId());
+
+			const update =
+				columnSelection === 'multiple'
+					? handleMultiSelection(allColIds, colId, context.current.selectedColumns, event.event)
+					: [colId];
+			setSelectedColumns(update);
+		},
+		[setSelectedColumns, columnSelection],
+	);
+
+	// Ag-grid doesn't provide click listener on header cell, create one
+	const enrichedColumnDefs = useMemo(() => {
+		return columnDefs?.map((colDef: ColDef) => ({
+			...colDef,
+			headerComponentParams: {
+				...colDef.headerComponentParams,
+				onFocus: onHeaderFocus,
+			},
+		}));
+	}, [columnDefs, onHeaderFocus]);
+
+	// Handle controlled selection
+	const controlledColumnIds = selection?.columnIds;
+	useEffect(() => {
+		if (controlledColumnIds) {
+			setSelectedColumns(controlledColumnIds);
+		}
+	}, [controlledColumnIds]);
+
+	// Force style update when selection changed
+	useEffect(() => {
+		const previousSelection = context.current.selectedColumns;
+		context.current.selectedColumns = selectedColumns;
+
+		// BEGIN QUICKFIX: Would be better to use gridRef.current?.api.refreshHeader() but it has a weird flicker
```

Comment:
What if you set the display of the grid (through opacity, not display: none) to a timer? You keep a skeleton running above for the duration of that timer. Fade out the skeleton / fade in the grid with refreshHeader?

Reply:
Discussed IRL, that's not it.